File: backend/app/harness/pipeline/runner.py
"""
PhaseRunner — topologically ordered phase executor.

Responsibilities:
- Validate the phase DAG at construction (Kahn topo sort + cycle detection).
- For each phase in order:
    1. Check Answer Now interrupt flag (if phase opts in).
    2. Publish SSE round_status (start of progress_range).
    3. Fire PRE_<NAME> hook if mapped to an existing HookPoint.
    4. Run phase.execute(ctx). Catch PhaseSkipped silently.
    5. Fire POST_<NAME> hook if mapped.
    6. Publish SSE round_status (end of progress_range).
- On PhaseAborted: re-raise so caller can return early without marking failed.
- On any other Exception: log and re-raise (caller marks round failed).
"""
from __future__ import annotations

import logging
from collections import defaultdict, deque
from typing import Optional

from app.harness.hook_engine import HookEngine, HookPoint
from app.services.event_bus import EventBus

from .types import Phase, PhaseAborted, PhaseSkipped, RoundContext

logger = logging.getLogger(__name__)


# Map phase.name → (PRE hook, POST hook). Phases without a mapping skip hooks.
# Keeps backwards compatibility with existing HookPoint enum without forcing a
# new value per phase.
_HOOK_MAP: dict[str, tuple[Optional[HookPoint], Optional[HookPoint]]] = {
    "load_round": (HookPoint.ROUND_START, None),
    "fetch": (None, HookPoint.POST_SEARCH),
    "score": (HookPoint.PRE_SCORING, HookPoint.POST_SCORING),
    "dispatch_summaries": (HookPoint.PRE_SUMMARIZE, None),
}


class PhaseRunner:
    def __init__(self, phases: list[Phase]):
        self._phases = self._topo_sort(phases)
        self._hooks = HookEngine.get_instance()

    @staticmethod
    def _topo_sort(phases: list[Phase]) -> list[Phase]:
        if not phases:
            return []
        by_name: dict[str, Phase] = {}
        for p in phases:
            if p.name in by_name:
                raise ValueError(f"duplicate phase name: '{p.name}'")
            by_name[p.name] = p

        indeg: dict[str, int] = {p.name: 0 for p in phases}
        graph: dict[str, list[str]] = defaultdict(list)
        for p in phases:
            for d in p.deps:
                if d not in by_name:
                    raise ValueError(
                        f"phase '{p.name}' depends on unknown phase '{d}'"
                    )
                graph[d].append(p.name)
                indeg[p.name] += 1

        ready: deque[str] = deque(n for n, k in indeg.items() if k == 0)
        order: list[Phase] = []
        while ready:
            n = ready.popleft()
            order.append(by_name[n])
            for m in graph[n]:
                indeg[m] -= 1
                if indeg[m] == 0:
                    ready.append(m)
        if len(order) != len(phases):
            unresolved = [n for n, k in indeg.items() if k > 0]
            raise ValueError(f"cycle detected involving phases: {unresolved}")
        return order

    @property
    def phases(self) -> list[Phase]:
        return list(self._phases)

    async def run(self, ctx: RoundContext) -> RoundContext:
        for p in self._phases:
            if await self._maybe_partial(ctx, p):
                # Partial answer delivered; abort the round.
                raise PhaseAborted(
                    "user_requested_partial",
                    payload={"partial": True, "stage": _partial_stage(p)},
                )

            # 声明式 conditional skip: phase.skip_if(ctx) → True 时整段跳过
            # (比 phase 内部 raise PhaseSkipped 更易写测试 + 不用先进入 hook)
            if await self._should_skip(p, ctx):
                logger.info("[phase] %s skipped via skip_if", p.name)
                ctx.artifacts[p.name] = None
                self._publish_progress(ctx.round_id, p, edge="start")
                self._publish_progress(ctx.round_id, p, edge="end")
                continue

            self._publish_progress(ctx.round_id, p, edge="start")

            pre_hook, post_hook = _HOOK_MAP.get(p.name, (None, None))
            if pre_hook is not None:
                await self._fire(pre_hook, _hook_ctx(ctx, p))

            try:
                output = await p.execute(ctx)
            except PhaseSkipped as e:
                logger.info("[phase] %s skipped: %s", p.name, e)
                ctx.artifacts[p.name] = None
                # Still emit end-progress so UI doesn't get stuck.
                self._publish_progress(ctx.round_id, p, edge="end")
                continue

            ctx.artifacts[p.name] = output

            if post_hook is not None:
                await self._fire(post_hook, _hook_ctx(ctx, p))

            self._publish_progress(ctx.round_id, p, edge="end")
        return ctx

    @staticmethod
    async def _should_skip(phase: Phase, ctx: RoundContext) -> bool:
        """Honor optional `phase.skip_if(ctx) -> bool | Awaitable[bool]`.

        允许同步 / 异步实现; 缺失或抛异常都视作"不跳过", 让 phase 正常执行.
        """
        skip_if = getattr(phase, "skip_if", None)
        if skip_if is None:
            return False
        try:
            result = skip_if(ctx)
            if hasattr(result, "__await__"):
                result = await result
            return bool(result)
        except Exception as e:
            logger.warning(
                "[phase] %s.skip_if() raised %s; running phase as fallback",
                phase.name, e,
            )
            return False

    async def _maybe_partial(self, ctx: RoundContext, phase: Phase) -> bool:
        if not getattr(phase, "can_interrupt", False):
            return False
        from app.services.partial_synthesizer import (
            clear_interrupt_flag,
            is_interrupt_requested,
            synthesize_partial,
        )
        from app.services.answer_now_helpers import deliver_partial_answer

        if not await is_interrupt_requested(ctx.round_id, ctx.redis):
            return False
        try:
            docs_so_far = _collect_docs_so_far(ctx)
            partial = await synthesize_partial(
                round_id=ctx.round_id,
                project_description=(
                    ctx.project.description if ctx.project else ""
                ),
                docs_so_far=docs_so_far,
                current_stage=_partial_stage(phase),
                llm_manager=ctx.llm_manager,
            )
            await deliver_partial_answer(ctx.round_id, partial, ctx.db)
        finally:
            await clear_interrupt_flag(ctx.round_id, ctx.redis)
        return True

    async def _fire(self, hook: HookPoint, payload: dict) -> None:
        try:
            await self._hooks.fire(hook, payload)
        except Exception as e:
            logger.warning("[hook] %s failed: %s", hook.value, e)

    @staticmethod
    def _publish_progress(round_id: str, phase: Phase, edge: str) -> None:
        start, end = phase.progress_range
        progress = start if edge == "start" else end
        EventBus.publish_sync(round_id, "round_status", {
            "status": phase.name,
            "progress": round(progress, 3),
            "message": f"{phase.name} {edge}",
        })


def _partial_stage(phase: Phase) -> str:
    """UI expects legacy stage names ('searching'/'scoring'/'saving'/'summarizing').
    Phase can override via class attr partial_stage; fall back to name."""
    return getattr(phase, "partial_stage", None) or phase.name


def _hook_ctx(ctx: RoundContext, phase: Phase) -> dict:
    """Build a minimal hook payload from RoundContext + phase, mirroring what
    the legacy code passed (round_id, project_id, etc.)."""
    payload: dict = {
        "round_id": ctx.round_id,
        "phase": phase.name,
    }
    if ctx.project is not None:
        payload["project_id"] = str(ctx.project.id)
        if hasattr(ctx.project, "description"):
            payload["project_description"] = (ctx.project.description or "")[:200]
    if ctx.round is not None and hasattr(ctx.round, "round_number"):
        payload["round_number"] = ctx.round.round_number
    return payload


def _collect_docs_so_far(ctx: RoundContext) -> list:
    """Best-effort: prefer score output, then rerank, then fetch."""
    for key in ("score", "rerank", "fetch"):
        if ctx.has(key):
            out = ctx.artifacts[key]
            if isinstance(out, dict):
                docs = out.get("selected_docs") or out.get("docs")
                if docs:
                    return docs
    return []

File: client/src/data/pipeline/runner.ts
/**
 * PhaseRunner — 客户端版，移植自 backend `app/harness/pipeline/runner.py`。
 *
 * 差异：
 * - 删除 Answer Now `can_interrupt` 检查（PRD 决策 10）
 * - 删除 HookEngine fire（hook_engine 是 backend 概念）
 * - 删除 SSE EventBus.publish_sync；progress 事件通过 ctx.eventBus 发到客户端 in-memory bus
 * - 用 AbortSignal 替代 Answer Now：phase 间检查 abortSignal.aborted 决定是否中断
 *
 * 行为：
 * 1. 构造时拓扑排序（Kahn 算法 + cycle detection）
 * 2. 依次执行每个 phase
 * 3. phase.skipIf(ctx) → true 直接跳过（不抛异常更声明式）
 * 4. PhaseSkipped 异常等价于 skipIf=true
 * 5. PhaseAborted 异常 → re-throw 让 caller 决定是否标 failed
 * 6. 其它异常 → re-throw 让 caller 标 failed
 * 7. 每个 phase 开始/结束 emit `round_status` 事件（progress 取自 progressRange）
 */
import {
  PhaseAborted,
  PhaseSkipped,
  type RoundContext,
} from './context'

export interface Phase {
  readonly name: string
  readonly deps: readonly string[]
  /** [start, end] ∈ [0, 1]，phase 开始/结束时分别 publish progress。 */
  readonly progressRange: readonly [number, number]
  /** 可选：声明式跳过条件，runner 在 execute 前调（同步或异步皆可）。 */
  skipIf?: (ctx: RoundContext) => boolean | Promise<boolean>
  /** 主逻辑。返回值会被 runner 写到 `ctx.artifacts[name]`。 */
  execute: (ctx: RoundContext) => Promise<unknown>
}

export class PhaseRunner {
  private readonly _phases: ReadonlyArray<Phase>

  constructor(phases: ReadonlyArray<Phase>) {
    this._phases = PhaseRunner._topoSort(phases)
  }

  get phases(): ReadonlyArray<Phase> {
    return this._phases
  }

  /** Kahn 拓扑排序，含 cycle / unknown-dep / duplicate-name 检测。 */
  private static _topoSort(phases: ReadonlyArray<Phase>): ReadonlyArray<Phase> {
    if (phases.length === 0) return []

    const byName = new Map<string, Phase>()
    for (const p of phases) {
      if (byName.has(p.name)) {
        throw new Error(`duplicate phase name: '${p.name}'`)
      }
      byName.set(p.name, p)
    }

    const indeg = new Map<string, number>()
    const graph = new Map<string, string[]>()
    for (const p of phases) {
      indeg.set(p.name, 0)
      graph.set(p.name, [])
    }
    for (const p of phases) {
      for (const d of p.deps) {
        if (!byName.has(d)) {
          throw new Error(`phase '${p.name}' depends on unknown phase '${d}'`)
        }
        graph.get(d)!.push(p.name)
        indeg.set(p.name, (indeg.get(p.name) ?? 0) + 1)
      }
    }

    const ready: string[] = []
    for (const [n, k] of indeg.entries()) if (k === 0) ready.push(n)

    const order: Phase[] = []
    while (ready.length > 0) {
      const n = ready.shift()!
      order.push(byName.get(n)!)
      for (const m of graph.get(n) ?? []) {
        const next = (indeg.get(m) ?? 0) - 1
        indeg.set(m, next)
        if (next === 0) ready.push(m)
      }
    }
    if (order.length !== phases.length) {
      const unresolved = [...indeg.entries()]
        .filter(([, k]) => k > 0)
        .map(([n]) => n)
      throw new Error(`cycle detected involving phases: [${unresolved.join(', ')}]`)
    }
    return order
  }

  /**
   * 顺序执行所有 phase。
   *
   * @throws PhaseAborted — runner 不吞，caller 决定是否标 failed
   * @throws Error — 任何 phase 抛非 PhaseSkipped 异常
   */
  async run(ctx: RoundContext): Promise<RoundContext> {
    for (const p of this._phases) {
      if (ctx.abortSignal.aborted) {
        throw new PhaseAborted('user_cancelled', { stage: p.name })
      }

      // 声明式跳过
      if (await PhaseRunner._shouldSkip(p, ctx)) {
        ctx.set(p.name, null)
        this._publishProgress(ctx, p, 'start')
        this._publishProgress(ctx, p, 'end')
        continue
      }

      this._publishProgress(ctx, p, 'start')

      let output: unknown
      try {
        output = await p.execute(ctx)
      } catch (e) {
        if (e instanceof PhaseSkipped) {
          ctx.set(p.name, null)
          this._publishProgress(ctx, p, 'end')
          continue
        }
        // PhaseAborted / 其它异常 → re-throw
        throw e
      }

      ctx.set(p.name, output)
      this._publishProgress(ctx, p, 'end')
    }
    return ctx
  }

  private static async _shouldSkip(phase: Phase, ctx: RoundContext): Promise<boolean> {
    if (!phase.skipIf) return false
    try {
      const r = phase.skipIf(ctx)
      if (r && typeof (r as Promise<boolean>).then === 'function') {
        return Boolean(await r)
      }
      return Boolean(r)
    } catch (e) {
      console.warn(
        `[PhaseRunner] ${phase.name}.skipIf() raised ${(e as Error).message}; running phase as fallback`,
      )
      return false
    }
  }

  private _publishProgress(ctx: RoundContext, phase: Phase, edge: 'start' | 'end'): void {
    const [start, end] = phase.progressRange
    const progress = edge === 'start' ? start : end
    try {
      ctx.eventBus.publish(`round:${ctx.roundId}`, 'round_status', {
        roundId: ctx.roundId,
        status: phase.name,
        progress: Math.round(progress * 1000) / 1000,
        message: `${phase.name} ${edge}`,
      })
    } catch (e) {
      // event 发不出来不能挂主流程
      console.warn('[PhaseRunner] eventBus.publish failed:', e)
    }
  }
}
